mbuild/compound.py

```diff
@@ -2388,6 +2384,212 @@ def _iterate_children(self, nodes, edges, names_only=False):
+        infer_residues : bool, optional, default=False
+            Attempt to assign residues based on names of children
+
+        Returns
+        ------
+        pybel.Molecule
+
+        Notes
+        -----
+        Most of the mb.Compound is first converted to openbabel.OBMol 
+        And then pybel creates a pybel.Molecule from the OBMol
+        Bond orders are assumed to be 1
+        OBMol atom indexing starts at 1, with spatial dimension Angstrom
+        """
+
+        openbabel = import_('openbabel')
+        pybel = import_('pybel')
+
+        mol = openbabel.OBMol()
+        particle_to_atom_index = {}
+
+        if not residues and infer_residues:
+            residues = list(set([child.name for child in self.children]))
+        if isinstance(residues, string_types):
+            residues = [residues]
+        if isinstance(residues, (list, set)):
+            residues = tuple(residues)
+
+        compound_residue_map = dict()
+        atom_residue_map = dict()
+
+        for i, part in enumerate(self.particles(include_ports=show_ports)):
+            if residues and part.name in residues:
+                current_residue = mol.NewResidue()
+                current_residue.SetName(part.name)
+                atom_residue_map[part] = current_residue
+                compound_residue_map[part] = current_residue
+            elif residues:
+                for parent in part.ancestors():
+                    if residues and parent.name in residues:
+                        if parent not in compound_residue_map:
+                            current_residue = mol.NewResidue()
+                            current_residue.SetName(parent.name)
+                            compound_residue_map[parent] = current_residue
+                        atom_residue_map[part] = current_residue
+                        break
+                else:  # Did not find specified residues in ancestors.
+                    current_residue = mol.NewResidue()
+                    current_residue.SetName("RES")
+                    atom_residue_map[part] = current_residue
+            else:
+                current_residue = mol.NewResidue()
+                current_residue.SetName("RES")
+                atom_residue_map[part] = current_residue
+
+            temp = mol.NewAtom()
+            residue = atom_residue_map[part]
+            temp.SetResidue(residue)
+            if part.port_particle:
+                temp.SetAtomicNum(0)
+            else:
+                try:
+                    temp.SetAtomicNum(AtomicNum[part.name.capitalize()])
+                except KeyError:
+                    warn("Could not infer atomic number from "
+                            "{}, setting to 0".format(part.name))
+                    temp.SetAtomicNum(0)
+
+
+            temp.SetVector(*(part.xyz[0]*10))
+            particle_to_atom_index[part] = i
+
+        ucell = openbabel.OBUnitCell()
+        if box is None:
```

Comment:
One thing I will comment on is that we are now using this code in a few places (lattice and here at least). We probably want to move this into a separate function to minimize code duplication.

But this should not be a target at the moment for this PR.